src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs

```diff
@@ -333,7 +348,97 @@ private async Task<Expression> RecurseChainedExpression(ChainedExpression expres
+
+                        queryPartitionStatistics = _queryPartitionStatisticsCache.GetQueryPartitionStatistics(searchOptions.Expression);
+                        if (IsQuerySelective(queryPartitionStatistics))
+                        {
+                            feedOptions.MaxConcurrency = MaxQueryConcurrency;
+                        }
+
+                        // plant a ConcurrentBag int the request context's properties, so the CosmosResponseProcessor
+                        // knows to add the individual ResponseMessages sent as part of this search.
+
+                        fhirRequestContext = _requestContextAccessor.FhirRequestContext;
+                        if (fhirRequestContext != null)
+                        {
+                            messagesList = new ConcurrentBag<ResponseMessage>();
+                            fhirRequestContext.Properties[Constants.CosmosDbResponseMessagesProperty] = messagesList;
+                        }
+                    }
+                }
+            }
+
+            try
+            {
+                var result = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(sqlQuerySpec, feedOptions, continuationToken, searchOptions.MaxItemCountSpecifiedByClient, cancellationToken);
+
+                if (queryPartitionStatistics != null && messagesList != null)
+                {
+                    // determine the number of unique physical partitions queried as part of this search.
+                    int physicalPartitionCount = messagesList.Select(r => r.Headers["x-ms-documentdb-partitionkeyrangeid"]).Distinct().Count();
+
+                    queryPartitionStatistics.Update(physicalPartitionCount);
+                }
+
+                return result;
+            }
+            finally
+            {
+                if (queryPartitionStatistics != null && fhirRequestContext != null)
+                {
+                    fhirRequestContext.Properties.Remove(Constants.CosmosDbResponseMessagesProperty);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Heuristic for determining whether a query is selective or not. If it is, we should query partitions in parallel.
+        /// If it is not, we should query sequentially, since we would expect to get a full page of results from the first partition.
+        /// This is really simple right now
+        /// </summary>
+        private bool IsQuerySelective(QueryPartitionStatistics queryPartitionStatistics)
+        {
+            int? averagePartitionCount = queryPartitionStatistics.GetAveragePartitionCount();
+
+            if (averagePartitionCount.HasValue)
+            {
+                // this is not a new query
+
+                double fractionOfPartitionsHit = (double)averagePartitionCount.Value / _physicalPartitionInfo.PhysicalPartitionCount;
+
+                if (fractionOfPartitionsHit > 0.5)
```

Comment:
Just clarifying, does this mean that on average, the query was executed across more than 50% of the physical partitions?

Reply:
Yes, that's right.

Reply:
More precisely, on average, we needed more than 50% of the physical partitions to fill a page of results.